enhancements/workload-partitioning/wide-availability-workload-partitioning.md

```diff
@@ -0,0 +1,407 @@
+1. The
+   [PerformanceProfile](https://github.com/openshift/cluster-node-tuning-operator/blob/master/docs/performanceprofile/performance_profile.md)
+   part of [Cluster Node Tuning
+   Operator](https://github.com/openshift/cluster-node-tuning-operator)
+
+We want to remove the checks in the admission controller which specifically
+checks that partitioning is only applied to single node topology configuration.
+The design and configuration for any pod modification will remain the same, we
+simply will allow you to apply partitioning on non single node topologies.
+
+Workload pinning involves configuring CRI-O and Kubelet. Currently, this is done
+through a machine config that contains both of those configurations. This can
+pose problems as the CPU set value has to be copied to these other two
+configurations. We want to simplify the current implementation and apply both of
+these configurations via the `PerformanceProfile` CRD.
+
+We want to add a new `Workloads` field to the `CPU` field that contains the
+configuration information for `enablePinning`. We are not sure where we would
+want to take workload pinning in the future and to allow flexibility we want to
+place the configuration in `cpu.workloads`.
+
+```yaml
+apiVersion: performance.openshift.io/v2
+kind: PerformanceProfile
+metadata:
+  name: openshift-node-workload-partitioning-custom
+spec:
+  cpu:
+    isolated: 2-3
+    reserved: 0,1
+    # New addition
+    workloads:
+      enablePinning: true
+```
+
+### Workflow Description
+
+The end user will be expected to provide a `PerformanceProfile` manifest that
+describes their desired `isolated` and `reserved` CPUSet and the
+`workloads.enablePinning` flag set to true. This manifest will be applied during
+the installation process.
+
+**High level sequence diagram:**
+
+```mermaid
+sequenceDiagram
+Alice->>Installer: Provide PerformanceProfile manifest
+Installer-->>NTO: Apply
+NTO-->>MCO: Generated Machine Manifests
+MCO-->>Node: Configure node
+loop Apply
+    Node->>Node: Set kubelet config
+    Node->>Node: Set crio config
+    Node->>Node: Kubelet advertises cores
+end
+Node-->>MCO: Finished Restart
+MCO-->>NTO: Machine Manifests Applied
+NTO-->>Installer: PerformanceProfile Applied
+Installer-->>Alice: Cluster is Up!
+```
+
+- **Alice** is a human user who creates an Openshift cluster.
+- **Installer** is assisted installer that applies the user manifest.
+- **NTO** is the node tuning operator.
+- **MCO** is the machine config operator.
+- **Node** is the kubernetes node.
+
+1. Alice sits down and provides the desired performance profile as an extra
+   manifest to the installer.
+1. The installer applies the manifest.
+1. The NTO will generate the appropriate machine configs that include the
+   kubelet config and the crio config to be applied as well as the existing
+   operation.
+1. Once the MCO applies the configs, the node is restarted and the cluster
+   installation continues to completion.
```

Comment:
I haven't tested this yet, but if my understanding is correct, since it will all be packaged into one machine config manifest we should see similar or possibly less restarts.

Reply:
The pao controller consolidates the various mc fragments into a single machine config so there should not be any additional reboots due to this functionality

Reply:
Just for a quick test, was able to spin up a cluster with the changes we are proposing, the restart count stayed the same between the existing way and the proposed way.